### PR TITLE
"pin verify".

### PR DESCRIPTION
**Pull request by kevina**

### CODE REVIEW

**Comment by kevina** [comment]
This needs work, but want to ge tthe U.I down first.  Here is the current usage:

```
$ ipfs pin verify
QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ok
QmVLDAhCY3X9P2uRudKAryuQFPM5zqA3Yij1dY8FpGbL7T ok
QmQZ1WVkddYPpRPm6vPpemhNacjwhupzSMzFEcYMisVpuY broken
  QmSijovevteoY63Uj1uC5b8pkpDU5Jgyk2dYBqz3sMJUPc: merkledag: not found
  QmTbPEyrA1JyGUHFvmtx1FNZVzdBreMv8Hc8jV9sBRWhNA: The block referred to by 'QmTbPEyrA1JyGUHFvmtx1FNZVzdBreMv8Hc8jV9sBRWhNA' was not a vald merkledag node
```

File: core/commands/pin.go [hunk 1]
```diff
@@ -414,6 +436,90 @@ func pinLsAll(typeStr string, ctx context.Context, n *core.IpfsNode) (map[string
 	return keys, nil
 }
 
+type pinStatus struct {
+	// use pointer to array slice to reduce memory usage
```
Comment (whyrusleeping) [inline core/commands/pin.go]: How does this reduce memory usage?

Comment (whyrusleeping) [inline core/commands/pin.go]: Just make it a normal slice. Slices are already pointers

File: core/commands/pin.go [hunk 2]
```diff
@@ -414,6 +436,90 @@ func pinLsAll(typeStr string, ctx context.Context, n *core.IpfsNode) (map[string
...
+		for _, cid := range recPins {
+			pinStatus := checkPin(cid)
+			out <- pinVerifyRes{cid, pinStatus}
+		}
+	}()
+
+	return out
+}
+
+func (r pinVerifyRes) Format(out io.Writer) {
```
Comment (whyrusleeping) [inline core/commands/pin.go]: Use a marshaler for this.

File: core/commands/pin.go [hunk 3]
```diff
@@ -332,6 +333,27 @@ Example:
...
+		out := pinVerify(req.Context(), n)
+
+		go func() {
+			defer wtr.Close()
+			for r := range out {
+				r.Format(wtr)
+			}
+		}()
+
+		res.SetOutput(rdr)
```
Comment (whyrusleeping) [inline core/commands/pin.go]: try not to ever set the output to a reader, that makes the http api realllllly hard to deal with

Comment (kevina) [inline core/commands/pin.go]: @whyrusleeping this is a WIP, more concerned with the U.I.  will fix up this later.

File: core/commands/pin.go (hunk 1 above)
Comment (kevina) [inline core/commands/pin.go]: A slice is the width of three pointers.  There is a node for every indirect pin.  By using an extra lawyer of indirection I reduce the size by two pointers for each node.  The actual array slice will be shared many times in the common case.

**Comment by kevina** [comment]
Here is the current usage:
```
USAGE
  ipfs pin verify - Verify that recursive pins are complete.

SYNOPSIS
  ipfs pin verify [--verbose] [--quiet | -q]

OPTIONS

  --verbose        bool - Also write the hashes of non-broken pins.
  -q,      --quiet bool - Write just hashes of broken pins.

```

Here as some sample outputs:
```
$ ipfs pin verify
QmQZ1WVkddYPpRPm6vPpemhNacjwhupzSMzFEcYMisVpuY broken
  QmSijovevteoY63Uj1uC5b8pkpDU5Jgyk2dYBqz3sMJUPc: merkledag: not found
  QmTbPEyrA1JyGUHFvmtx1FNZVzdBreMv8Hc8jV9sBRWhNA: The block referred to by 'QmTbPEyrA1JyGUHFvmtx1FNZVzdBreMv8Hc8jV9sBRWhNA' was not a vald merkledag node
```
```
$ ipfs pin verify -q
mQZ1WVkddYPpRPm6vPpemhNacjwhupzSMzFEcYMisVpuY
```
```
$ ipfs pin verify --verbose
QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ok
QmVLDAhCY3X9P2uRudKAryuQFPM5zqA3Yij1dY8FpGbL7T ok
QmQZ1WVkddYPpRPm6vPpemhNacjwhupzSMzFEcYMisVpuY broken
  QmSijovevteoY63Uj1uC5b8pkpDU5Jgyk2dYBqz3sMJUPc: merkledag: not found
  QmTbPEyrA1JyGUHFvmtx1FNZVzdBreMv8Hc8jV9sBRWhNA: The block referred to by 'QmTbPEyrA1JyGUHFvmtx1FNZVzdBreMv8Hc8jV9sBRWhNA' was not a vald merkledag node
```

@whyrusleeping @Kubuxu (and others) let me know if you think the usage and output is reasonable and I will work on adding test cases and fixing the API.

**Comment by kevina** [comment]
@whyrusleeping @Kubuxu  (and others) let me know if you think the usage and output is reasonable and I will work on adding test cases and fixing the API.

**Comment by hsanjuan** [comment]
Is this a client side operation only, not exposed through http api?

**Comment by kevina** [comment]
@hsanjuan what makes you say that?

**Comment by hsanjuan** [comment]
maybe just ignorance @kevina. Just not clear to me what the output would be in JSON since structs have private fields.

**Comment by kevina** [comment]
> Just not clear to me what the output would be in JSON since structs have private fields.
@hsanjuan that part is not designed yet

At this point I just want to know "if you think the usage and output is reasonable".

This is not ready for a full review yet.

**Comment by hsanjuan** [comment]
@kevina yeah, so this would involve knowing if the JSON output is reasonable, and how it matches the CLI-formatted output etc..

**Comment by kevina** [comment]
I was hoping for some feedback on the U.I. before I concern myself with the API and the JSON output.

**Comment by whyrusleeping** [comment]
Yeah, given that the output looks nontrivial (nested things and different sorts of behaviour) it would probably be good to spec out what the api looks like here as that may be just as important as the output.

That said, I think the given output looks good to me. It provides decent enough information to the end user

**Comment by kevina** [comment]
@hsanjuan @whyrusleeping I fixed the API.

**Comment by hsanjuan** [comment]
@kevina thanks, can you post the json output for the examples above?

**Comment by kevina** [comment]
I would if I knew @hsanjuan an easy easy way to get at it.  I will hand construct it later today if necessary.

**Comment by Kubuxu** [comment]
@kevina `ipfs --enc=json pin verify`.

**Comment by kevina** [comment]
Standard output:
```
> ipfs --enc=json pin verify
{"Cid":"zdvgq3v8bUfT3EskMzK7FqzpMiymfmTLfZA8JZh43wr9RMk64","Ok":false,"BadNodes":[{"Cid":"zdvgq3v8bUfT3EskMzK7FqzpMiymfmTLfZA8JZh43wr9RMk64","Err":"unrecognized object type: %!s(uint64=114)"}]}
...
{"Cid":"QmRKmNZqwSxcPs6HjEN3JiAKxjEJMByYpgr5ykwKuqNahP","Ok":false,"BadNodes":[{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"},{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"},{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"},{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"}]}
```
With `--quiet` option.
```
> ipfs --enc=json pin verify -q
{"Cid":"zdvgqDYeRDUp9EFkTwt4MDH4GFsZKu5wAA2x5zxcPYxGNZks6","Ok":false}
{"Cid":"QmRKmNZqwSxcPs6HjEN3JiAKxjEJMByYpgr5ykwKuqNahP","Ok":false}
```
With `--verbose option.
```
{"Cid":"QmYAUTWVubcjipibdrwkJPbH61XNDrERHyAABRqwWU9K46","Ok":true}
...
{"Cid":"zdvgq3v8bUfT3EskMzK7FqzpMiymfmTLfZA8JZh43wr9RMk64","Ok":false,"BadNodes":[{"Cid":"zdvgq3v8bUfT3EskMzK7FqzpMiymfmTLfZA8JZh43wr9RMk64","Err":"unrecognized object type: %!s(uint64=114)"}]}
...
{"Cid":"QmRKmNZqwSxcPs6HjEN3JiAKxjEJMByYpgr5ykwKuqNahP","Ok":false,"BadNodes":[{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"},{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"},{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"},{"Cid":"zdvgqAiZHaP4GiRxrhecMdw6rb2wBMrP7r2CGbicCznebfLTH","Err":"unrecognized object type: %!s(uint64=114)"}]}
```

**Comment by whyrusleeping** [comment]
@kevina that api output looks good to me

**Comment by kevina** [comment]
@whyrusleeping rebased

**Pull request merged**